File: src/main/java/com/matag/game/turn/action/selection/SelectedPermanentsGetAction.java
package com.matag.game.turn.action.selection;

import org.springframework.stereotype.Component;

import com.matag.cards.ability.selector.SelectorType;
import com.matag.game.cardinstance.CardInstance;
import com.matag.cards.ability.AbilityAction;
import com.matag.cards.ability.Ability;
import com.matag.game.status.GameStatus;
import com.matag.game.turn.action.permanent.PermanentGetService;
import com.matag.game.turn.action.player.PlayerGetService;

import lombok.AllArgsConstructor;

@Component
@AllArgsConstructor
public class SelectedPermanentsGetAction implements AbilityAction {
  private final MagicInstancePermanentSelectorService magicInstancePermanentSelectorService;
  private final MagicInstancePlayerSelectorService magicInstancePlayerSelectorService;
  private final PermanentGetService permanentGetService;
  private final PlayerGetService playerGetService;

  @Override
  public void perform(CardInstance cardInstance, GameStatus gameStatus, Ability ability) {
    var magicInstanceSelector = ability.getMagicInstanceSelector();
    if (magicInstanceSelector.getSelectorType() == SelectorType.PLAYER) {
      magicInstancePlayerSelectorService.selectPlayers(gameStatus, cardInstance, magicInstanceSelector)
        .forEach(player -> playerGetService.thatPlayerGets(cardInstance, gameStatus, ability.getParameters(), player));

    } else {
      magicInstancePermanentSelectorService.select(gameStatus, cardInstance, magicInstanceSelector).getCards()
        .forEach(card -> permanentGetService.thatPermanentGets(cardInstance, gameStatus, ability.getParameters(), card));
    }
  }
}

File: src/main/kotlin/com/matag/game/turn/action/selection/SelectedPermanentsGetAction.kt
package com.matag.game.turn.action.selection

import com.matag.cards.ability.Ability
import com.matag.cards.ability.AbilityAction
import com.matag.cards.ability.selector.SelectorType
import com.matag.game.cardinstance.CardInstance
import com.matag.game.player.Player
import com.matag.game.status.GameStatus
import com.matag.game.turn.action.permanent.PermanentGetService
import com.matag.game.turn.action.player.PlayerGetService
import org.springframework.stereotype.Component
import java.util.function.Consumer

@Component
class SelectedPermanentsGetAction(
    private val magicInstancePermanentSelectorService: MagicInstancePermanentSelectorService,
    private val magicInstancePlayerSelectorService: MagicInstancePlayerSelectorService,
    private val permanentGetService: PermanentGetService,
    private val playerGetService: PlayerGetService
) : AbilityAction {
    
    override fun perform(cardInstance: CardInstance, gameStatus: GameStatus, ability: Ability?) {
        val magicInstanceSelector = ability!!.magicInstanceSelector
        if (magicInstanceSelector!!.selectorType == SelectorType.PLAYER) {
            magicInstancePlayerSelectorService.selectPlayers(gameStatus, cardInstance, magicInstanceSelector)
                .forEach(Consumer { player: Player? ->
                    playerGetService.thatPlayerGets(
                        cardInstance,
                        gameStatus,
                        ability.parameters,
                        player!!
                    )
                })
        } else {
            magicInstancePermanentSelectorService.select(gameStatus, cardInstance, magicInstanceSelector).cards
                .forEach(Consumer { card: CardInstance? ->
                    permanentGetService.thatPermanentGets(
                        cardInstance,
                        gameStatus,
                        ability.parameters,
                        card!!
                    )
                })
        }
    }
}
